### ISSUE TITLE
wgpu-py doesn't ship binaries on WSL (Windows Subsystem for Linux)

### ISSUE THREAD

**Issue by FirefoxMetzger** [pygfx/wgpu-py#292]
When installing wgpu on WSL the installed version doesn't ship any binaries. Subsequently, the tests fail (complaining that they can't find the binaries). I didn't see any necessary post-install steps in the README, but I could have of course missed something. Do I need to import them from somewhere manually? (on native windows the binaries were included in the package.)

For reference; here is the linux version I'm using: Linux DESKTOP-2SO47AE 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

**Comment by almarklein** [comment]
There is the [download-wgpu-native.py](https://github.com/pygfx/wgpu-py/blob/main/download-wgpu-native.py) script, but this is intended for people that checked out the repo; it is not available when installed via pip.

I would have expected pip on WSL to download the Linux wheel, but apparently this is not the case?

**Comment by Korijn** [comment]
Well, there are only a couple of reasons pip would not install the manylinux2010 wheel on your distribution:

* Your linux distribution does not comply with the manylinux2010 distribution (very unlikely)
* Your version of python does not comply with the python requirement (`>=3.7`)
* Your version of pip and wheel are not recent enough to be aware of the manylinux2010 tag

Please be aware that if you are going to work with WSL, you won't be able to use hardware acceleration/your GPU. There are currently no compatible drivers. You would have to use software rendering (lavapipe).

**Comment by almarklein** [comment]
Ok, so maybe a `pip install -U pip` would do the trick 😄

**Comment by FirefoxMetzger** [comment]
> Your version of pip and wheel are not recent enough to be aware of the manylinux tag used here

That does the trick. WSL ships with pip 20.0.2 by default, which is indeed not very recent. Interestingly, `pip` doesn't seem to warn that it is outdated, so that one seems hard to catch. Thanks for the pointer @Korijn !

This does bring up an interesting question though: Should the installation through pip fail if we can't detect a suitable pre-compiled wheel? At the moment pip falls back to building from source, which appears to succeed (but doesn't compile anything); however, it doesn't contain the binary which I'd argue is quite core to functioning correctly. Would it be better UX if this install failed instead? One way to do so would be to add a step to setup.py that checks if there is a binary present and fails otherwise. This way, we can still publish the sdist, but it won't install from pip. Alternatively, we could look into not releasing a sdist, which might achieve the same (would have to investigate that option).

@Korijn @almarklein If this sounds like a useful idea, we could take it into a new issue and close this one?

> Please be aware that if you are going to work with WSL, you won't be able to use hardware acceleration/your GPU. There are currently no compatible drivers. You would have to use software rendering (lavapipe).

Hm, I thought there is some GPU support via [WSLg](https://github.com/microsoft/wslg). I've used this successfully in the past to run a robotics simulator and it was definitely doing something on the GPU. I did, however, just learn that it only supports OpenGL, so I guess it's of not so much use in our scenario.

**Comment by almarklein** [comment]
> That does the trick. WSL ships with pip 20.0.2 by default, which is indeed not very recent. 

From what I see in the [pip changelog](https://pip.pypa.io/en/stable/news/#v20-3), we'd need pip version 20.3 for it to recognize our manylinux_2_24. Would be good to also mention in the installation instructions that pip 20.3+ is needed.

I just tried wgpu on WSL (on Windows 11) again. At first I got segfaults, but after `apt upgrade` the triangle example worked fine! Looks like a more recent version of lavapipe fixed that. I will create an issue that we should also add a section on WSL (and other Linuxes without a Vulkan context) in the docs.

**Comment by almarklein** [comment]
I commented before responding to all points ...

>  Should the installation through pip fail if we can't detect a suitable pre-compiled wheel? 

No, because sometimes it may be intentional, or someone could be on a very special platform that is not supported by wheels. These users can somehow get their own copy of wgpu-native (by compiling it themselves, for instance) and then use `WGPU_LIB_PATH`.

Nevertheless,  we could emit a useful warning when we're not able to load the binary. Maybe even when the sdist is installed instead of a wheel?

I created some issues that came out of this question:
* #296 
* #297 
* #298
